# 02 · Numbers

**Learning goals**

- Use Python as a calculator with the arithmetic operators.
- Tell `/`, `//` and `%` apart.
- Understand operator precedence and when parentheses help.
- Know what to expect from very large numbers, floats and division by zero.

> In a notebook the value of the **last expression** in a cell is displayed
> automatically. Use `print()` when you want to show something else.

## Integers and basic arithmetic

Python evaluates arithmetic the way you would on paper: `+` adds, `-`
subtracts and `*` multiplies. Negative numbers work as you would expect.

In [1]:
-456 + 654

198

In [2]:
8 * 14

112

`print()` writes text to the output area. Unlike a bare expression, it shows
the value without quotes and can print several things in a row.

In [3]:
print("Hiii")

Hiii


## Division: three different operators

| Operator | Name | Example | Result |
| :--- | :--- | :--- | :--- |
| `/` | true division (always a `float`) | `15 / 4` | `3.75` |
| `//` | floor division (drops the fraction) | `15 // 4` | `3` |
| `%` | remainder (modulo) | `15 % 4` | `3` |

Notice that `/` returns a `float` even when the division is exact.

In [4]:
20 / 5

4.0

In [5]:
15 / 4

3.75

In [6]:
15 // 4

3

In [7]:
15 % 4

3

`//` rounds **down** (toward negative infinity), not toward zero. This only
shows up with negative numbers:

In [8]:
-7 // 2

-4

## Remainder tricks

A number is divisible by `k` exactly when `n % k == 0`. Checking `n % 2 == 0`
is the standard test for an even number.

In [9]:
957 % 3

0

In [10]:
print(957 % 3 == 0)  # divisible by 3?
print(376 % 2 == 0)  # even?

True
True


## Powers and very large numbers

`**` is the power operator. Python integers have no fixed size limit, so
results that would overflow in many other languages are still exact.

In [11]:
7**4

2401

In [12]:
34**45

825184889837725994971915471173621692744697928074424045623011467853824

## Floats are approximations

Numbers with a decimal point are stored as binary floating-point values, so
many decimal fractions cannot be represented exactly.

In [13]:
87.26 / 42.8

2.038785046728972

In [14]:
0.1 + 0.2

0.30000000000000004

## Readable numbers and operator precedence

Underscores in a number literal are ignored, which makes long numbers easier
to read. Multiplication and division are evaluated before addition and
subtraction; use parentheses to make the order explicit.

In [15]:
9_999_999

9999999

In [16]:
2 + 3 * 4

14

In [17]:
(2 + 3) * 4

20

In [18]:
((4 + 3) * (8 / 2)) * 5

140.0

The last result is `140.0`, not `140`, because `8 / 2` produced a float.

## Errors are information: division by zero

Some operations are impossible. Python stops and raises an *exception*. Here
we **catch** it with `try` / `except` so the notebook keeps running and we can
read the message.

In [19]:
try:
    7 / 0
except ZeroDivisionError as error:
    print(f"ZeroDivisionError: {error}")

ZeroDivisionError: division by zero


## Key takeaways

- `/` gives a float, `//` gives the floored quotient, `%` gives the remainder.
- Use parentheses whenever the order of operations is not obvious.
- Integers are exact at any size; floats are approximations.
- Dividing by zero raises `ZeroDivisionError`.

**Next:** [03 · Strings](03_strings.ipynb)